flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/AvroToProtoSerializer.java

```diff
@@ -297,9 +328,344 @@ private static Object handleLogicalTypeSchema(Schema fieldSchema, Object value)
+            mapping.put("geography_wkt", AvroSchemaHandler::convertGeography);
+            mapping.put("Json", AvroSchemaHandler::convertJson);
             return mapping.get(logicalTypeString);
         }
+
+        // ---- Utilities to enable Conversions for Logical Types ------------
+        @VisibleForTesting
+        static String convertUUID(Object value) {
+            if (value instanceof UUID) {
+                return ((UUID) value).toString();
+            } else if (value instanceof String) {
+                UUID.fromString((String) value);
+                return (String) value;
+            }
+            LOG.error(getLogErrorMessage("String/UUID", "UUID", value.getClass().toString()));
+            throw new IllegalArgumentException(getErrorMessage("String/UUID", "UUID"));
+        }
+
+        private static void validateTimestamp(long timestamp) {
+            try {
+                // Validate timestamp from microseconds since EPOCH.
+                Timestamp.ofTimeMicroseconds(timestamp);
+            } catch (IllegalArgumentException e) {
+                LOG.error(
+                        String.format(
+                                "Invalid Timestamp '%s' Provided."
+                                        + "\nShould be a long value indicating microseconds since "
+                                        + "Epoch (1970-01-01 00:00:00.000000+00:00) between %s and %s",
+                                timestamp, Timestamp.MIN_VALUE, Timestamp.MAX_VALUE));
+                throw new IllegalArgumentException(
+                        String.format("Invalid Timestamp '%s' Provided.", timestamp));
+            }
+        }
+
+        // BigQuery inputs timestamp as microseconds since EPOCH,
+        // So if we have TIMESTAMP in micros - we convert as it is.
+        // If the TIMESTAMP is in millis - we convert to Micros and then add.
+        @VisibleForTesting
+        static Long convertTimestamp(Object value, boolean micros, String type) {
+            long timestamp;
+            if (value instanceof ReadableInstant) {
+                timestamp = TimeUnit.MILLISECONDS.toMicros(((ReadableInstant) value).getMillis());
+            } else if (value instanceof Long) {
+                timestamp = (micros ? (Long) value : TimeUnit.MILLISECONDS.toMicros((Long) value));
+            } else {
+                LOG.error(
+                        getLogErrorMessage(
+                                "LONG/ReadableInstant",
+                                "TIMESTAMP (" + type + ")",
+                                value.getClass().toString()));
+                throw new IllegalArgumentException(
+                        getErrorMessage("LONG/ReadableInstant", "TIMESTAMP"));
+            }
+            validateTimestamp(timestamp);
+            return timestamp;
+        }
+
+        private static void validateDate(Integer date) {
+            if (date > 2932896 || date < -719162) {
```

Comment:
Lets define constants with appropriate names for these apparent max and min date values.

Reply:
Added!